libsplinter/src/network/peer_manager/mod.rs

```diff
@@ -0,0 +1,1040 @@
+                continue;
+            }
+        }
+    }
+
+    warn!("Unable to peer with {}", peer_id);
+    // remove the reference created above
+    ref_map.remove_ref(&peer_id);
+    // unable to connect to any of the endpoints provided
+    Err(PeerRefAddError::AddError(format!(
+        "Unable to connect any endpoint that was provided for peer {}",
+        peer_id
+    )))
+}
+
+fn update_peer(
+    peer_id: String,
+    new_peer_id: String,
+    peers: &mut PeerMap,
+    ref_map: &mut RefMap,
+) -> Result<(), PeerRefUpdateError> {
+    // update the ref_map, so old PeerRef can still be used to drop references
+    if ref_map
+        .update_ref(peer_id.clone(), new_peer_id.clone())
+        .is_err()
+    {
+        return Err(PeerRefUpdateError::UpdateError(format!(
+            "Unable to update peer, {} does not exist",
+            peer_id
+        )));
+    }
+
+    // update the peer in the peer map
+    match peers.update_peer_id(peer_id.clone(), new_peer_id.clone()) {
+        Ok(()) => {
+            debug!("Updated peer id from {} to {}", peer_id, new_peer_id);
+            Ok(())
+        }
+        Err(_) => Err(PeerRefUpdateError::UpdateError(format!(
+            "Unable to update peer, {} does not exist",
+            peer_id
+        ))),
+    }
+}
+
+fn remove_peer(
+    peer_id: String,
+    connector: Connector,
+    peers: &mut PeerMap,
+    ref_map: &mut RefMap,
+) -> Result<(), PeerRefRemoveError> {
+    // remove the reference
+    let removed_peer = ref_map.remove_ref(&peer_id);
+
+    if let Some(removed_peer) = removed_peer {
+        let active_endpoint = peers.remove(&removed_peer).ok_or_else(|| {
+            PeerRefRemoveError::RemoveError(format!(
+                "Peer {} has already been removed from the peer map",
+                peer_id
+            ))
+        })?;
+
+        match connector.remove_connection(&active_endpoint) {
+            Ok(Some(_)) => {
+                debug!(
+                    "Peer {} has been removed and connection {} has been closed",
+                    peer_id, active_endpoint
+                );
+                Ok(())
+            }
+            Ok(None) => Err(PeerRefRemoveError::RemoveError(
+                "No connection to remove, something has gone wrong".to_string(),
+            )),
+            Err(err) => Err(PeerRefRemoveError::RemoveError(format!("{}", err))),
+        }
+    } else {
+        // if the peer has not been fully removed, return OK
+        Ok(())
+    }
+}
+
+// If a connection has reached the retry limit before it could be reestablished, the peer manager
+// will try the peer's other endpoints.
+fn retry_endpoints(
+    peer_id: &str,
+    connector: Connector,
+    peers: &mut PeerMap,
+) -> Result<bool, PeerManagerError> {
+    let peer_metadata = match peers.peer_metadata(peer_id) {
+        Some(peer_metadata) => peer_metadata.clone(),
+        // this should never happen
+        None => {
```

Comment:
Why not pass the metadata in and return the update to the peer map, if successful. That way, we don't have to event modify the PeerMap in this function.

Reply:
i will update

Reply:
Updated, also add PeerStatus enum and removed disconnected/connenctioned methods on the PeerMap.